pinot-controller/src/main/java/org/apache/pinot/controller/LeadControllerManager.java

```diff
@@ -0,0 +1,157 @@
+    int partitionIndex =
+        HelixHelper.getPartitionIdForTable(rawTableName, NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE);
+    if (_leadControllerChecker.isPartitionLeader(partitionIndex)) {
+      return true;
+    } else {
+      return isHelixLeader();
+    }
+  }
+
+  /**
+   * Get lead controller id and partition index for given table. Returns -1 as partition index and null as lead controller id if not found.
+   * @param tableName table name with/without table type.
+   * @return lead controller id and partition index
+   */
+  public LeadControllerResponse getLeadControllerAndPartitionIdForTable(String tableName) {
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    ExternalView leadControllerResourceExternalView = _pinotHelixResourceManager.getHelixAdmin()
+        .getResourceExternalView(_pinotHelixResourceManager.getHelixClusterName(), LEAD_CONTROLLER_RESOURCE_NAME);
+    if (leadControllerResourceExternalView == null) {
+      return new LeadControllerResponse();
+    }
+    Set<String> partitionSet = leadControllerResourceExternalView.getPartitionSet();
+    if (partitionSet == null || partitionSet.isEmpty()) {
+      return new LeadControllerResponse();
+    }
+
+    int numPartitions = partitionSet.size();
+    int partitionIndex = HelixHelper.getPartitionIdForTable(rawTableName, numPartitions);
+    String leadControllerId = HelixHelper.getLeadControllerForTable(leadControllerResourceExternalView, rawTableName);
+    return new LeadControllerResponse(partitionIndex, leadControllerId);
+  }
+
+  public synchronized void addPartitionLeader(String partitionName) {
+    _leadControllerChecker.addPartitionLeader(partitionName);
+  }
+
+  public synchronized void removePartitionLeader(String partitionName) {
+    _leadControllerChecker.removePartitionLeader(partitionName);
+  }
+
+  /**
+   * Checks if the current controller host is Helix leader.
+   */
+  private boolean isHelixLeader() {
+    return _amIHelixLeader;
+  }
+
+  /**
+   * Response class to return lead controller id and partition index.
+   */
+  public class LeadControllerResponse {
+    private int _partitionIndex;
+    private String _leadControllerId;
+
+    public LeadControllerResponse(int partitionIndex, String leadControllerId) {
+      _partitionIndex = partitionIndex;
+      _leadControllerId = leadControllerId;
+    }
+
+    public LeadControllerResponse() {
+      _partitionIndex = -1;
+      _leadControllerId = null;
+    }
+
+    public int getPartitionIndex() {
+      return _partitionIndex;
+    }
+
+    public String getLeadControllerId() {
+      return _leadControllerId;
+    }
+
+    @Override
+    public String toString() {
+      return "Partition index: " + getPartitionIndex() + ", lead controller id: " + getLeadControllerId();
+    }
+  }
+
+  /**
+   * Callback on changes in the controller. Should be registered to the controller callback.
```

Comment:
can you elaborate this comment? This callback will only be triggered if the controller is in dual mode, right? It won't be triggered in any other mode.

Reply:
Nope, the participant resource manager will register the controller listener. So once there's a controller change, all the participant will get notified, i.e. it'll work in both dual-mode and in Pinot-only mode.

Reply:
Please add a comment that this callback will be invoked whether we are in dual mode or not. It helps to clarify it. You can also probably repeat the comment in ControllerStarter that this is here to support the case when the lead controller resource is disabled, and is otherwise unused